packages/openapi-v3/src/consolidate-schema.ts

```diff
@@ -0,0 +1,106 @@
+import compare from 'json-schema-compare';
+import _ from 'lodash';
+import {
+  ISpecificationExtension,
+  isSchemaObject,
+  OpenApiSpec,
+  PathsObject,
+  ReferenceObject,
+  SchemaObject,
+} from './types';
+
+let $schemaRefs: {
+  [schema: string]: SchemaObject | ReferenceObject;
+};
+let $paths: PathsObject;
+
+/*
+ *  Recursively search OpenApiSpec PathsObject for SchemaObjects with title property.
+ *  Move reusable schema bodies to #/components/schemas and replace with json pointer.
+ *  Handles collisions /w title, schema body pair comparision.
+ */
+export function ConsolidateSchemaObjects(spec: OpenApiSpec): OpenApiSpec {
+  $schemaRefs =
+    spec.components && spec.components.schemas
+      ? _.cloneDeep(spec.components.schemas)
+      : {};
+  $paths = _.cloneDeep(spec.paths);
+
+  recursiveWalk($paths);
+
+  const updatedSpec = {
+    ...spec,
+    ...{
+      paths: $paths,
+      components: {...spec.components, ...{schemas: $schemaRefs}},
```

Comment:
IIUC, this code replaces `spec.components.schemas` with `$schemaRefs`. Is that a safe operation? The name `schemaRefs` implies to me that this object contains only references (schemas extracted by our algorithm?), but what about the original component schemas provided by the user?

Reply:
> but what about the original component schemas provided by the user?

Original schemas in `spec.components.schemas` cloned to `$schemaRefs` at start.  `$schemaRefs` used as working object then replaces original.  Would it help to rename anything?